rfd/0163-vnet.md

```diff
@@ -0,0 +1,796 @@
+Not to mention that the user's private key is stored at `~/.tsh` and can be used
+to access anything the user can access as well.
+
+VNet is purely a client-side convenience the effectively automates what you can
+already do by creating TCP proxies for all apps.
+It is built on top of Teleport's security principals like short-lived
+certificates, device trust, passwordless authentication, and MFA.
+
+VNet does not add much extra attack surface unless the user's laptop is already
+compromised.
+
+### UX
+
+VNet is available in both tsh and Teleport Connect. This section focuses mostly on VNet integration
+in Teleport Connect, with a short subsection on VNet in tsh at the end.
+
+There are three goals that the VNet integration attempts to achieve.
+
+1. VNet should stay out of the way.
+1. VNet should serve as a complete replacement for app gateways that shipped with Connect v15.
+1. The user should be naturally guided towards enabling VNet whenever they perform an action that
+   needs VNet to be running.
+
+Before we show how the UI for VNet works, there's an important requirement we need to address.
+
+#### Root privileges
+
+As described in [Host configuration](#host-configuration), for VNet to work we need to create a TUN
+interface and configure DNS for a particular cluster. Both actions require root privileges.
+
+The first version of VNet uses osascript's `do shell script … with administrator privileges` to
+call a tsh subcommand as root. This means that VNet doesn't work in environments where the user is a
+standard macOS user and cannot elevate their privileges.
+
+While [osascript is a CLI tool intended to be used by sysadmins and not as an
+API](https://forums.developer.apple.com/forums/thread/701118?answerId=705661022#705661022), it gets
+us past the problem of acquiring root privileges and lets us focus on validating VNet as a tool.
+In the [Alternatives to osascript](#alternatives-to-osascript) section, we describe other approaches
+which offer better UX at the cost of development time.
+
+#### Guiding the user towards VNet
+
+Before VNet, the user would access web and TCP apps through the unified resources view in Connect:
+
+<img width="871" alt="Unified Resources view in Connect v15.0.0" src="assets/0163-apps-in-connect-v15.png">
+
+Clicking "Launch" next to a web app would launch this app in the browser through the `/web/launch`
+URL of the proxy service. The user could also set up a local proxy for the specific app by clicking
+on the three dots menu and choosing "Set up connection". For TCP apps, clicking "Connect" would set
+up a local proxy for that app.
+
+This changes with the introduction of VNet. Here's a screenshot from Teleport Connect in cluster
+`teleport-local.dev` running VNet:
+
+<img width="871" alt="Unified Resources view in Connect with VNet turned on" src="assets/0163-apps-with-vnet.png">
+
+Web apps now have two actions available in the three dots menu: "Connect and copy VNet address" and
```

Comment:
I'm going to change it to just "Copy address" since there will be no separate VNet address.

Reply:
I guess "Copy VNet address" is still relevant here because for web apps, we want the VNet address to be the HTTP version of public addr.